### PR TITLE
Init-container templates

### PR DESCRIPTION
**Pull request by DTadrzak**
PR relates to #69 and #89 

Adding: templates for init-containers. Unfortunately I need to break it for
two templates, because i don't how to pass two variables to template.

PR also adjust layout of configuration files in Nova.

Signed-off-by: DTadrzak <daniel.tadrzak@intel.com>

---
This change is [Reviewable](https://reviewable.io/reviews/att-comdev/openstack-helm/104)

### CODE REVIEW

File: cinder/templates/deployment-api.yaml [hunk 1]
```diff
@@ -19,31 +19,8 @@ spec:
...
-                "value": "{{  include "joinListWithColon" .Values.dependencies.api.jobs }}"
-              },
-              {
-                "name": "COMMAND",
-                "value": "echo done"
-              }
-            ]
-          }
-        ]'
+      {{  include "init-containers-header" . }}
```
Comment (alanmeadows) [inline cinder/templates/deployment-api.yaml]: To collapse this you could potentially follow the pattern of "template":

https://github.com/att-comdev/openstack-helm/blob/master/common/templates/_funcs.tpl#L5-L7

Using the tuple method to pass in the arguments referenced with index:

https://github.com/att-comdev/openstack-helm/blob/92512d66bb45d078775c92d324dbedb85a686536/neutron/templates/configmap-etc.yaml#L7

I don't know of a better way - I have expressed to @technosophos on #helm that being able to provide input to template "blocks" needs to be more intuitive.

File: common/templates/_funcs.tpl [hunk 2]
```diff
@@ -21,3 +21,42 @@
 {{- $wtf := $context.Template.Name | replace $last $name -}}
 {{- include $wtf $context | sha256sum | quote -}}
 {{- end -}}
+
+{{- define "init-containers-header"}}
```
Comment (alanmeadows) [inline common/templates/_funcs.tpl]: I worry this precludes people from adding "chart specific" init containers to the charts - if you shifted the template to start at ```{ "name" ....}``` they could definite additional init containers that are not common?

Comment (alanmeadows) [inline common/templates/_funcs.tpl]: We may also want to be specific with this define name, its specifically a dependency check init container.

**Comment by intlabs** [comment]
Hey, this looks like a good start, but I think needs a few changes:
- Should be reworked to be a single template (you may find the stuff I did for endpoint registration useful here)
- The template should only contain the init container itself, allowing us to add additional init containers to a pod as required. This can be a bit tricky with the current json implementation, but I think having to restrict the init container to the last element is an acceptable compromise for now.
- Obviously, I appreciate that this is a WIP, but the development helper scripts should be removed.

File: common/templates/_funcs.tpl [hunk 3]
```diff
@@ -22,37 +22,34 @@
...
-      }
-    ]
-  }
-]'
+{{- define "dep-check-init-cont" -}}
+{{- $envALL := index . 0 -}}
+{{- $deps := index . 1 -}}
+{
+  "name": "init",
+  "image": {{ $envALL.Values.images.dep_check | default "quay.io/stackanetes/kubernetes-entrypoint:v0.1.0" | quote }},
```
Comment (intlabs) [inline common/templates/_funcs.tpl]: We shouldn't have an image declared as a default.

File: cinder/templates/deployment-api.yaml [hunk 4]
```diff
@@ -19,8 +21,9 @@ spec:
       annotations:
         configmap-bin-hash: {{ tuple "configmap-bin.yaml" . | include "hash" }}
         configmap-etc-hash: {{ tuple "configmap-etc.yaml" . | include "hash" }}
-{{  include "dep-check-init-cont-header" . | indent 8 }}
-{{  include "dep-check-init-cont-footer" .Values.dependencies.api | indent 8 }}
+        pod.beta.kubernetes.io/init-containers: '[
+{{ tuple $envAll $dependecies | include "dep-check-init-cont" | indent 10 }}
```
Comment (intlabs) [inline cinder/templates/deployment-api.yaml]: Typo, this affects all instances:
s/dependecies/dependencies

File: common/templates/_funcs.tpl [hunk 5]
```diff
@@ -22,37 +22,34 @@
...
+  "imagePullPolicy": {{ $envALL.Values.images.pull_policy | default "IfNotPresent" | quote }},
```
Comment (intlabs) [inline common/templates/_funcs.tpl]: I'm a bit on the fence here, but my gut says that we should not have default values in templates at all if possible.

File: common/templates/_funcs.tpl [hunk 6]
```diff
@@ -22,37 +22,34 @@
...
+  "env": [
+    {
+      "name": "NAMESPACE",
+      "value": "{{ $envALL.Release.Namespace }}"
+    },
+    {
+      "name": "INTERFACE_NAME",
+      "value": "eth0"
+    },
+    {
```
Comment (intlabs) [inline common/templates/_funcs.tpl]: Should also include downward API env values and ability to check daemonsets.

File: heat/templates/deployment-cfn.yaml [hunk 7]
```diff
@@ -1,3 +1,5 @@
+{{- $envAll := . }}
+{{- $dependecies := .Values.dependencies.cnf }}
```
Comment (intlabs) [inline heat/templates/deployment-cfn.yaml]: Typo:
s/cnf/cfn

File: common/templates/_funcs.tpl [hunk 8]
```diff
@@ -22,37 +22,34 @@
...
-      },
-      {
-        "name": "COMMAND",
-        "value": "echo done"
-      }
-    ]
-  }
-]'
+{{- define "dep-check-init-cont" -}}
+{{- $envALL := index . 0 -}}
```
Comment (intlabs) [inline common/templates/_funcs.tpl]: I'd keep this as $envAll to maintain consistency

File: common/templates/_funcs.tpl [hunk 9]
```diff
@@ -22,37 +22,34 @@
...
-        "value": "{{  include "joinListWithColon" .jobs }}"
-      },
-      {
-        "name": "COMMAND",
-        "value": "echo done"
-      }
-    ]
-  }
-]'
+{{- define "dep-check-init-cont" -}}
```
Comment (intlabs) [inline common/templates/_funcs.tpl]: For consistency could we used '_' instead of '-' as a delineator?

Also rather than overloading the _funcs.tpl I'd be tempted to move this into its own file within the `snippets` directory.

File: common/templates/_funcs.tpl (hunk 6 above)
Comment (DTadrzak) [inline common/templates/_funcs.tpl]: Let's move it to another PR that will cover this and introduce templatable init containers to Neutron.

File: heat/templates/deployment-cfn.yaml (hunk 7 above)
Comment (DTadrzak) [inline heat/templates/deployment-cfn.yaml]: Good Catch :)

File: common/templates/_funcs.tpl (hunk 5 above)
Comment (DTadrzak) [inline common/templates/_funcs.tpl]: Yeaha you are right.

File: common/templates/_funcs.tpl (hunk 6 above)
Comment (alanmeadows) [inline common/templates/_funcs.tpl]: Agreed, and note POD_NAME will be required for daemonset checks.  Fine with this in another commit so we can get this in.

**Comment by v1k0d3n** [comment]
All reviews LG.
